### ISSUE TITLE
Strange predictions

### ISSUE THREAD

**Issue by kevaday** [xgeric/definition-embedding#1]
Hello,
I trained the dictionary model on the felix dictionary with the following arguments:
-hidden size: 512
-# of layers: 16
-batch size: 128
-lr: 0.001
-optimizer: adam
I allowed it to train for a while and achieved a loss of about 0.8, then I modified the case.ipynb to predict sentences. The code is as follows:
```
import torch
import pickle
import argparse
from os.path import abspath

import sys
sys.path.append(abspath('../src'))
import model.util as util
from model.rnn import RNNEncoder as rnn
from model.bow import BOWEncoder as bow


def main(args):
    # load checkpoint
    checkpoint = torch.load(f'../checkpoint/{args.checkpoint}', map_location=lambda storage, loc: storage)

    # load data
    model_args = checkpoint['args']
    with open(model_args.data, 'rb') as rf:
        data = pickle.load(rf)
    dic_embed = data['dic_embed']
    def_embed = data['def_embed']
    dic_word2ix = data['dic_word2ix']
    def_word2ix = data['def_word2ix']
    dic_ix2word = {v: k for k, v in dic_word2ix.items()}
    rvd_candidates = set(data['rvd_candidates'])
    print('[1] load data OK:', model_args.data)

    # load model
    model_args.gpu = args.gpu
    encoder = eval(args.model)(def_word2ix, model_args)
    if args.gpu > -1:
        encoder.cuda()
    encoder.load_state_dict(checkpoint['state_dict'])
    print('[2] load model OK!')

    dic_embed = util.normalize_matrix_by_row(dic_embed.numpy())

    def rank_sen(sen):
        out_embs = encoder.estimate_from_strsens([sen], normalize=True)
        simi_matrix = out_embs.dot(dic_embed.transpose()).reshape(-1)
        rank = sorted(range(len(dic_word2ix)), key=lambda i: simi_matrix[i], reverse=True)

        return [dic_ix2word[i] for i in rank[:20]]
    
    for s in args.sentences:
        print(rank_sen(s))


def get_args():
    parser = argparse.ArgumentParser('predict phrases')
    parser.add_argument('-c', '--checkpoint', type=str, required=True, help='trained model checkpoint file path')
    parser.add_argument('-m', '--model', type=str, default='rnn', help='model type: rnn or bow')
    parser.add_argument('-g', '--gpu', type=int, default=-1, help='which gpu to use, -1 is cpu')
    parser.add_argument('-s', '--sentences', type=str, action='append', required=True, help='sentences to predict (use -s multiple times for more than one sentence)')
    
    return parser.parse_args()


if __name__ == '__main__':
    main(get_args())
```
The problem is that for every single prediction, the top 20 ranked words are:
`['sort', 'kind', 'rather', 'so', 'way', 'it', 'sometimes', 'you', 'little', 'this', 'simple', 'something', 'fashioned', 'like', 'sense', 'bit', 'even', 'much', 'too', 'turn']`
every single time. I cannot seem to figure out why this is happening. Help would be much appreciated!

Thanks

**Comment by xgeric** [comment]
Hi,

Thank you for reaching out. What are your testing sentences? It seems that the input indices to the model are the same every time. I would suggest you to set a break point and check the input indices (the "sen" list at https://github.com/xgeric/definition-embedding/blob/master/src/model/base.py#L183), and see if the input sentences are correctly converted to expected token indices. If not (e.g. all the input tokens are "UNK") then the model may get the same input for each sentence and give same output as well.

Just my guess. Let me know if you figure it out or still have questions : )

**Comment by kevaday** [comment]
Hi,

Thank you for the reply! Just making sure, `sen` is a string with a sentence such as "restricted to a particular place" and sens is a list of all the sentences to predict? Or am I getting it wrong somehow...

Thanks

**Comment by xgeric** [comment]
Please forgive my bad choice of variable naming orz... In line 183 of base.py, **sen** is a list of token ids of one tokenized string sentence, for instance, [1, 2, 3, ...]. **sens** is a list of **sen**. Your input looks good to me so this should not be the problem.

Just in case, do you keep getting the same output word list for the same sentence, or are you getting the same output list for different sentences?

**Comment by kevaday** [comment]
Oh yea forgive me for describing the situation incorrectly, I meant str_sens is a list of character sentences that are tokenized and appended to sens. I was not sure if the str_sens is a word, and the network is providing the definition, or str_sens is a sentence and the network should provide the word to suit it, but know I understand that the is the reverse task where the network provides the word to suit a definition.

I get the same output word list for any input that I give the network and I have tried multiple different sentences.

**Comment by xgeric** [comment]
I see. Yes this task is given a definition sentence and output a ranked word list, so your input should be good. Can you print the sens variable after line 184 of base.py for different input sentences?

**Comment by kevaday** [comment]
Using the sentence `"restricted to a particular place"`, after line 184 in base.py, sens is: `[[17346, 17026, 26366, 17587, 21897]]`

**Comment by xgeric** [comment]
OK, can you try a different sentence and see the output?

**Comment by kevaday** [comment]
The tokenized sentence (`sens`) for `"a male parent"` is `[[26366, 20218, 20888]]`

**Comment by xgeric** [comment]
All look good to me. This is weird. I'll try to check the code these days (as I haven't run this for a year). Another thing I notice is the number of layers you are currently using is large compared with the size of the dataset. Maybe you can try to run with the default hyper-parameters I specified in train.py to see if you get better performance. I don't quite remember if 0.8 is a normal loss but I'll check. Let me know if you have any further findings. Will get back to you once I rerun the code with your input : )

**Comment by kevaday** [comment]
All right thank you, I will try to retrain the model with the default hyperparameters. I used so many layers because I wanted to try to create a network that has quite a deep understanding of dictionary definitions and so that it can be flexible with definitions that it has not seen before. Anyways, have you trained the model with the default hyperparameters and decent results? If so, what loss it achieve and was it capable of defining unseen data?
Also, how would it be possible to run the network on the default task, so giving it a word would make it say the definition?

**Comment by xgeric** [comment]
I don't remember the loss achieved with default parameters but the evaluation metrics of the reverse dictionary task is close to the [original paper](https://arxiv.org/abs/1504.00548).

For definition generation from word embeddings I would recommend [another work](https://www.aaai.org/ocs/index.php/AAAI/AAAI17/paper/viewPDFInterstitial/14827/14211). They also published code I remember and that should be what you are looking for : )

**Comment by kevaday** [comment]
Alright thanks, I'll check it out!

**Comment by kevaday** [comment]
Do you know the link to the code by any chance?

**Comment by xgeric** [comment]
Yes you can find it on his homepage: https://github.com/Websail-NU/torch-defseq
